src/core/quantum_gate.jl

```diff
@@ -101,6 +126,124 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
+applies it. 
+
+# Examples
+Whereas a `Hadamard` Gate with `target=1` is constructed by calling `hadamard(target)`, a 
+controlled Hadamard gate is constructed using the `hadamard()` `Function`. For instance, the construction 
+of a `ControlledGate{Hadamard}` with `control=1` and `target=2` is performed by calling:
+
+```jldoctest controlled_hadamard
+julia> controlled_hadamard=ControlledGate(hadamard,1,2)
+Gate Object: ControlledGate{Snowflake.Hadamard}
+Connected_qubits	: [1, 2]
+Operator:
+(4, 4)-element Snowflake.DenseOperator:
+Underlying data ComplexF64:
+1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    0.7071067811865475 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    -0.7071067811865475 + 0.0im
+
+```
+
+It can then be used in a `QuantumCircuit` as any other `Gate`, and its display symbol is 
+inherited from the display symbol of the single-target `Hadamard` `Gate`:
+
+```jldoctest controlled_hadamard
+julia> circuit=QuantumCircuit(qubit_count=2,gates=[controlled_hadamard])
+Quantum Circuit Object:
+   qubit_count: 2 
+q[1]:──*──
+       |  
+q[2]:──H──
+          
+```
+
+In general, a `ControlledGate` with an arbitraty number of targets and controls can be 
+constructed. For instance, the following constructs the equivalent of a `Toffoli` `Gate`, 
+but as a `ConnectedGate{SigmaX}`, with `control_qubits=[1,2]` and `target_qubit=[3]`:
+
+```jldoctest 
+julia> toffoli_as_controlled_gate=ControlledGate(sigma_x,[1,2],[3])
+Gate Object: ControlledGate{Snowflake.SigmaX}
+Connected_qubits	: [1, 2, 3]
+Operator:
+(8, 8)-element Snowflake.DenseOperator:
+Underlying data ComplexF64:
+1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im
+```
+
+
+"""
+struct ControlledGate{GateType<:AbstractGate}<:AbstractGate
+    kernel::GateType
+    connected_qubits::Vector{Int}
+    control_qubits::Vector{Int}
+    kernel_qubits::Vector{Int}
+
+    function ControlledGate(
+        kernel_handle::Function,
```

Comment:
Don't take a function that constructs a gate. Take an already constructed gate. This limits what the user can define as their own gate.

Reply:
ok, though this opens up the recursive ControlledGate{ControlledGate{...}}. Should I put safeguards against that?

Reply:
updated, with assertion error for ControlledGate{ControlledGate}